### ISSUE TITLE
SSL pre-training and fine-tuning on `(64, 64, 64)` samples

### ISSUE THREAD

**Issue by valosekj** [ivadomed/model-seg-dcm#9]
This issue summarizes the follow-up experiment to https://github.com/ivadomed/model-seg-dcm/issues/7.

## Pre-training

In https://github.com/ivadomed/model-seg-dcm/issues/7., the entire input image of size [(64, 256, 256)](https://github.com/ivadomed/model-seg-dcm/blob/3bc999a116d447592677e01e981ea49a2d30fbfe/vit_unetr_ssl/train.py#L72C21-L72C33) was used to create two augmeted views. 
Here, two random samples of size [(64, 64, 64)](https://github.com/ivadomed/model-seg-dcm/blob/16691d77e72623898b8d9af024d4672f76e882e0/vit_unetr_ssl/train.py#L77C16-L77C28) are taken from the input image to create the augmented view using [these transforms](https://github.com/ivadomed/model-seg-dcm/blob/jv/vit_unetr_ssl/vit_unetr_ssl/transforms.py#L32-L133). Moreover, these samples are taken around a label (spinal cord mask) using [RandCropByPosNegLabeld](https://github.com/ivadomed/model-seg-dcm/blob/jv/vit_unetr_ssl/vit_unetr_ssl/transforms.py#L58-L66):

<img width="700" alt="image" src="https://github.com/ivadomed/model-seg-dcm/assets/39456460/aad43b3f-64c0-47fb-b350-38ef12993249">

**Comment by valosekj** [comment]
I tried three experiments, all for 500 epochs, modifying:
- training [batch_size](https://github.com/ivadomed/model-seg-dcm/blob/0e802dfb3495c68415a0872721a568cb92047235/vit_unetr_ssl/train.py#L120)
- [number of holes](https://github.com/ivadomed/model-seg-dcm/blob/0e802dfb3495c68415a0872721a568cb92047235/vit_unetr_ssl/transforms.py#L38) during samples augmentation 

## `batch_size=8`, `5 holes`

Reconstructed images across epochs: [link](https://www.dropbox.com/scl/fo/tkcs4fa9egzvmtbbbigt7/h?rlkey=3z1pi6kkutfc1m9k89j1lwb2g&dl=0)

👉 visually looks the best, training now for more epochs

<details><summary>loss_plots</summary>

![loss_plots](https://github.com/ivadomed/model-seg-dcm/assets/39456460/c593a5c5-0c5b-4105-9cd9-39e3901ecf97)

</details>

## `batch_size=8`, `10 holes`

Reconstructed images across epochs: [link](https://www.dropbox.com/scl/fo/7je79nkcsmfb5v62zmnvn/h?rlkey=v6xsj85409kdttybspci0d0q9&dl=0)

<details><summary>loss_plots</summary>

![loss_plots](https://github.com/ivadomed/model-seg-dcm/assets/39456460/fff6c1ec-66a2-48b8-b0bf-c888feb0bcad)

</details>

## `batch_size=16`, `10 holes`

Reconstructed images across epochs: [link](https://www.dropbox.com/scl/fo/tgy9ivj7f9n4adbllfnzk/h?rlkey=oihm3ejfy8ire72i91t8lqqxs&dl=0)

<details><summary>loss_plots</summary>

![loss_plots](https://github.com/ivadomed/model-seg-dcm/assets/39456460/02cc4c4f-a15a-459f-969a-210d105c7f93)

</details>

**Comment by valosekj** [comment]
Two experiments with [RandCoarseDropoutd augmentation transform](https://github.com/ivadomed/model-seg-dcm/blob/af9c6d09d2f57540675abff420e1d5b60acc055a/vit_unetr_ssl/transforms.py#L73-L124): `fill_value=None` vs `fill_value=0`. Discussion [here](https://github.com/ivadomed/model-seg-dcm/pull/10#discussion_r1527311715).
Both used [batch_size=8](https://github.com/ivadomed/model-seg-dcm/blob/af9c6d09d2f57540675abff420e1d5b60acc055a/vit_unetr_ssl/train.py#L121), [5 holes](https://github.com/ivadomed/model-seg-dcm/blob/af9c6d09d2f57540675abff420e1d5b60acc055a/vit_unetr_ssl/transforms.py#L34), and 1500 epochs. 

## `RandCoarseDropoutd fill_value=None`

= filling the dropped regions with randomly selected value based on the min and max value of input image

Reconstructed images across epochs: [link](https://www.dropbox.com/scl/fo/t6a7dtx6rvtouqbyl418l/h?rlkey=06snqfdxm43ggleye7r22bgks&dl=0)

<details><summary>loss_plots</summary>

![loss_plots](https://github.com/ivadomed/model-seg-dcm/assets/39456460/c436f5e5-493b-4ada-8ec6-a18d4dfae050)

</details>

## `RandCoarseDropoutd fill_value=0`

= filling the dropped regions with value `0`

Reconstructed images across epochs: [link](https://www.dropbox.com/scl/fo/8t7jvfdxxh6trvpglpm3h/h?rlkey=jufxw8mv30w8u9fkc89tide31&dl=0)

<details><summary>loss_plots</summary>

![loss_plots](https://github.com/ivadomed/model-seg-dcm/assets/39456460/c41136bd-2a71-4331-95cc-c73d36d88686)

</details>

**Comment by valosekj** [comment]
# Fine-tuning on `dcm-zurich-lesions`

Since the pre-training seems to work (at least the code does not raise errors and the model is learning to reconstruct the images), I performed two fine-tuning experiments: using a pre-trained model vs training from scratch.

## Pre-trained model

Fine-tuning using the SSL pre-trained model with `RandCoarseDropoutd fill_value=0` (see [comment above](https://github.com/ivadomed/model-seg-dcm/issues/9#issuecomment-2003705806)).

<details><summary>finetune loss plots  --> crashed to zero</summary>

![image](https://github.com/ivadomed/model-seg-dcm/assets/39456460/fe66bb7b-f67c-43f3-a8a7-f226ed57146f)

</details>

Although the learning crashed, the model was learning something! See the validation figure below.

<details><summary>validation figure</summary>

![image](https://github.com/ivadomed/model-seg-dcm/assets/39456460/7c045133-2130-47d6-8864-6a47778b8132)

</details>

## Training from scratch

Fine-tuning from the scratch (weights were initialized randomly).

<details><summary>loss plots  --> crashed to zero</summary>

![image](https://github.com/ivadomed/model-seg-dcm/assets/39456460/6f5f878e-8236-401c-868e-c81beb6343d7)

</details>
